### PR TITLE
std: Tweak expansion of thread-local const

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit tweaks the expansion of `thread_local!` when combined with a
`const { ... }` value to help ensure that the rules which apply to
`const { ... }` blocks will be the same as when they're stabilized.
Previously with this invocation:

    thread_local!(static NAME: Type = const { init_expr });

this would generate (on supporting platforms):

    #[thread_local]
    static NAME: Type = init_expr;

instead the macro now expands to:

    const INIT_EXPR: Type = init_expr;
    #[thread_local]
    static NAME: Type = INIT_EXPR;

with the hope that because `init_expr` is defined as a `const` item then
it's not accidentally allowing more behavior than if it were put into a
`static`. For example on the stabilization issue [this example][ex] now
gives the same error both ways.

[ex]: https://github.com/rust-lang/rust/issues/84223#issuecomment-953384298

### CODE REVIEW

File: library/std/src/thread/local.rs [hunk 1]
```diff
@@ -165,6 +165,7 @@ macro_rules! __thread_local_inner {
         #[cfg_attr(not(windows), inline)] // see comments below
         unsafe fn __getit() -> $crate::option::Option<&'static $t> {
             const _REQUIRE_UNSTABLE: () = $crate::thread::require_unstable_const_init_thread_local();
+            const INIT_EXPR: $t = $init;
```
Comment (nbdd0121) [inline library/std/src/thread/local.rs:168]: Probably `INIT_VAL` would be better, since it isn't an expression.

**Review by nbdd0121 — APPROVED** [review]

**Comment by hkratz** [comment]
Could this have caused the following aarch64 assertion failure in the rollup #90975 CI ([CI log](https://github.com/rust-lang-ci/rust/runs/4237469385?check_suite_focus=true#step:25:32065)):
```
fatal runtime error: assertion failed: thread_info.is_none()
```
stemming from:
https://github.com/rust-lang/rust/blob/11491938f80988c7261a1179cf71a25c379c8783/library/std/src/sys_common/thread_info.rs#L13
https://github.com/rust-lang/rust/blob/11491938f80988c7261a1179cf71a25c379c8783/library/std/src/sys_common/thread_info.rs#L41-L47

**Comment by alexcrichton** [comment]
@bors: r-

That seems likely, yes, I will work on investigating.

**Comment by alexcrichton** [comment]
@bors: r=m-ou-se rollup=never

Hm ok so I can't reproduce this locally and the IR looks the same before/after this change, so I'm going to put this back in the queue, but I'll exclude it from rollups in case it's still the culprit.

**Comment by alexcrichton** [comment]
It looks like dist-aarch64-apple took [over an hour](https://github.com/rust-lang-ci/rust/runs/4247698799?check_suite_focus=true#step:25:20175) to rebuild LLVM, so presumably caches were being primed.

@bors: retry

**Pull request merged**